Bump Ancient version to 0.10.0

Halbaroth opened the pull request:
A new version of Ancient have been released! This PR bumps the minimal
version of Ancient to 0.10.0.

There is no need to check the presence of `nnp` option or the operating
system in `configure.ml` as Ancient cannot be installed on Windows or
OCaml 4 without the nnp option.

Halbaroth committed:
Bump Ancient version to 0.10.0

A new version of Ancient have been released! This PR bumps the minimal
version of Ancient to 0.10.0.

There is no need to check the presence of `nnp` option or the operating
system in `configure.ml` as Ancient cannot be installed on Windows or
OCaml 4 without the nnp option.

Halbaroth committed:
Update flake.lock

We need to update the flake to download the latest commit on the branch
v0.10.x of ocaml-ancient.

Halbaroth committed:
Update opam in the dockerfile

Halbaroth committed:
Use ancient release in the CI

a2line commented:
Using `FROM ocaml/opam:debian-12-ocaml-4.14-nnp@sha256:114abd2ca7db4b9224f9c80b72d854f1d54c924c1574c229364d60af770a8364 AS builder` on a short specific Actions shows [it contains all the recent Opam releases, including latest 2.3.0](https://github.com/a2line/geneweb/actions/runs/15921336974/job/44908601407#step:3:42). So I gave it [a try](https://github.com/a2line/geneweb/actions/runs/15921658811/job/44909623484#step:6:1755) and it still failed:

``` bash
#18 278.1 <><> hkdf.1.0.4 installed successfully ><><><><><><><><><><><><><><><><><><><><>
#18 278.1 => Note: This package is deprecated.
#18 278.7 [ERROR] Package ancient has no version 0.10.0.
#18 ERROR: process "/bin/sh -c eval \"$(opam env)\"     && opam update     && opam install . --deps-only --with-test     && opam install ancient.0.10.0" did not complete successfully: exit code: 5
------
 > [builder 6/9] RUN eval "$(opam env)"     && opam update     && opam install . --deps-only --with-test     && opam install ancient.0.10.0:
210.2 -> installed tls.0.17.5
210.6 -> installed tls-lwt.0.17.5
211.0 -> installed gluten-lwt-unix.0.5.1
211.3 -> installed httpun-lwt-unix.0.2.0
274.9 -> installed camlp5.8.03.01
278.1 Done.
278.1 
278.1 <><> hkdf.1.0.4 installed successfully ><><><><><><><><><><><><><><><><><><><><>
278.1 => Note: This package is deprecated.
Error: ERROR] Package ancient has no version 0.10.0.`
```
The problem with Docker CI failing isn't updating opam itself (or redoing an opam update), I tested what repository was used by Docker that might be using a stale snapshot of the opam repository:
``` docker
#18 [builder 6/9] RUN eval "$(opam env)"     && echo "=== Debugging opam repositories ==="     && opam repository list -s     && echo "=== Searching for ancient packages ==="     && opam search ancient     && echo "=== Trying direct install first ==="     && (opam install ancient.0.10.0 || echo "Direct install failed, using git pin...")     && (opam list ancient || opam pin ancient -y https://github.com/OCamlPro/ocaml-ancient.git)     && opam install . --deps-only --with-test
#18 0.061 === Debugging opam repositories ===
#18 0.189 opam-repository-archive
#18 0.189 default
#18 0.192 === Searching for ancient packages ===
#18 0.938 # Packages matching: match(*ancient*)
#18 1.546 # Name  # Installed # Synopsis
#18 1.546 ancient --          Use data structures larger than available memory
#18 1.549 === Trying direct install first ===
#18 2.352 [ERROR] Package ancient has no version 0.10.0.
#18 2.355 Direct install failed, using git pin...
#18 2.983 # Packages matching: name-match(ancient) & (installed | available)
#18 3.586 # Package     # Installed # Synopsis
#18 3.586 ancient.0.9.0 --          Allows to use in-memory data structures which are larger than available memory and so are kept in swap
#18 3.586 ancient.0.9.1 --          Use data structures larger than available memory
```
The installation passed [using git pin](https://github.com/a2line/geneweb/actions/runs/15922018777/job/44910758423#step:6:1464), and so that's it: The Docker base image uses `opam-repository-archive` as Opam repository which is a cached version that doesn't include the recently published ancient! After a few other Docker CI trying to set that live repository, I have a [working workflow](https://github.com/a2line/geneweb/actions/runs/15923309485) keeping `-bare` at init followed by `set-url`:
``` dockerfile
FROM ocaml/opam:debian-12-ocaml-4.14-nnp@sha256:114abd2ca7db4b9224f9c80b72d854f1d54c924c1574c229364d60af770a8364 AS builder

ENV OPAMYES=yes
ENV OPAMJOBS=2
ENV DUNE_PROFILE=release

USER root
# Install required system dependencies
RUN export DEBIAN_FRONTEND=noninteractive \
    && apt-get update \
    && apt-get install -yq --no-install-recommends \
    m4 libgmp-dev libpcre3-dev libipc-system-simple-perl xdot zlib1g-dev pkg-config

# Use opam 2.3 and configure to use live repository like Colab
USER opam
WORKDIR /home/opam
RUN sudo ln -f /usr/bin/opam-2.3 /usr/bin/opam \
    && opam init --reinit --bare --disable-sandboxing -ni \
    && opam repository set-url default https://opam.ocaml.org

# Copy opam file for dependency resolution then install dependencies
COPY --chown=opam:opam *.opam ./
RUN eval "$(opam env)" \
    && opam install . --deps-only --with-test \
    && opam install ancient.0.10.0
```
So I think you can cherry-pick the last [commit on the Dockerfile](https://github.com/geneweb/geneweb/compare/master...a2line:geneweb:master) I used.

a2line commented:
Removed the SHA at start as it was for amd64 and it was showing a warning for arm64:
```[Base image platform does not match expected target platform: docker/Dockerfile#L5](https://github.com/a2line/geneweb/commit/08ce604340d6c3901e0d23fb6fb0315df5de9786#annotation_36125265801)
InvalidBaseImagePlatform: Base image ocaml/opam:debian-12-ocaml-4.14-nnp@sha256:114abd2ca7db4b9224f9c80b72d854f1d54c924c1574c229364d60af770a8364 was pulled with platform "linux/amd64", expected "linux/arm64" for current build
```

Halbaroth commented:
Thanks for your help. 

> The problem with Docker CI failing isn't updating opam itself (or redoing an opam update), I tested what repository was used by Docker that might be using a stale snapshot of the opam repository:

It was my fear that it would not be enough to fix the issue. It is better to use the latest opam release.

> The installation passed [using git pin](https://github.com/a2line/geneweb/actions/runs/15922018777/job/44910758423#step:6:1464), and so that's it: The Docker base image uses opam-repository-archive as Opam repository which is a cached version that doesn't include the recently published ancient

`opam-repository-archive` is not a cached version of `opam-repository`. It contains archived packages and it makes sense to add it in a docker to ensure that the docker still works in the future: https://github.com/ocaml/opam-repository-archive
We should have both opam-repository and opam-repository-archive.

Halbaroth commented:
I ran: 
```console
docker run --tty --interactive ocaml/opam:debian-12-ocaml-4.14-nnp@sha256:114abd2ca7db4b9224f9c80b72d854f1d54c924c1574c229364d60af770a8364
```
and `opam --version` in the prompt. I still got `2.0.10`. How did you get opam `2.3.0`?

The command `opam repository` outputs: 
```
 1 opam-repository-archive git+https://github.com/ocaml/opam-repository-archive
 2 default                 file:///home/opam/opam-repository
``` 
The image contains a copy of opam repository and we just need to update it.

Halbaroth commented:
Okay! There are both `opam 2.0.10` and `opam 2.3` in the image and you create a symbolic link in your docker file:
```sh 
ln -sf /usr/bin/opam-2.3 /usr/bin/opam
```

Halbaroth committed:
Update local opam repository in the docker file

The docker image contains a copy of the opam repository which is not up
to date. This commit changes dockerfile to pin this repository on a
specific commit.

Halbaroth commented:
It works :)

a2line commented:
I don't get why you pin a specific commit, is it better than HEAD on a maintenance perspective? (Edit: or `&& git checkout master`. Edit2 : or even simpler `git pull origin master` without fetching first!).

a2line commented:
You should remove 807dec36bb53e5e2418379cfeebe7b0f50077b07.

Halbaroth commented:
I already removed [807dec3](https://github.com/geneweb/geneweb/commit/807dec36bb53e5e2418379cfeebe7b0f50077b07) ! 

I need to run `opam update` after pulling the opam repository.

I've pinned both the docker image and the opam commit for sake of reproducible build. If I were to remove the pin on the opam repository, there would be no guarantee that dependencies of GeneWeb wouldn't be archived in the future and I would fetch the archive opam repository too.

a2line commented:
https://github.com/geneweb/geneweb/pull/2270/commits but I still see it!?

Note that I tried yesterday on latest Ubuntu LTS and `opam install ancient` does install the latest 0.10.0.

Halbaroth commented:
Oh you meant the commit itself. I think that we should squash all the PR before merging it. Its history is not relevant at all.

Did you run `opam update` before?

a2line commented:
I'm not sure about the local tests I did, I could have done it by habit. That said, I went to remove the version number afterward on the Colab script and it worked too and it doesn't have `opam update` I think it's not needed because it's already done at the `opam init` stage that is a line just before the installation of dependencies.

a2line commented:
I'm ok with squashing them and doing it, I tried a simple `opam install ancient` on both `Dockerfile` and `ci.yml`. Tests are passing and I verified it installed the latest 0.10.0 version as it should on both ([ci](https://github.com/a2line/geneweb/actions/runs/15970026612/job/45038957379#step:11:21) and [docker](https://github.com/a2line/geneweb/actions/runs/15970029248/job/45038966466#step:6:1641)). I also took care of squashing your nice comments to a single one, that is:
```
Bump Ancient version to 0.10.0

This PR bumps the minimal version of Ancient to new 0.10.0 release.

There is no need to check the presence of `nnp` option or the operating
system in `configure.ml` as Ancient cannot be installed on Windows or
OCaml 4 without that option.

- Update the flake to download the latest commit on the branch
v0.10.x of ocaml-ancient.
- Remove the pin to the Ancient repository in the CI and the Dockerfile.
- Update the Dockerfile local Opam repository by pinning a specific commit.
```

If you agree, I'll push that squashed commit to the master. It's very welcome, thank you again for that new Ancient release (isn't that sentence funny? ^^).

Halbaroth commented:
It's funny indeed :) you're welcome!

a2line commented:
squashed and merged in 0df69b4e8
